src/os/bluestore/BlueStore.cc

```diff
@@ -18513,12 +18554,359 @@ static uint32_t flush_extent_buffer_with_crc(BlueFS::FileWriter *p_handle, const
   return crc;
 }
 
-const unsigned MAX_EXTENTS_IN_BUFFER = 4 * 1024; // 4K extents = 64KB of data
+//-----------------------------------------------------------------------------------
+static void set_free_in_simple_bmap(CephContext* cct,
+  SimpleBitmap* sbmap, uint64_t offset, uint64_t length,
+  uint64_t mask, uint8_t order)
+{
+  dout(30) << "0x" << std::hex
+           << offset << "~" << length
+           << " " << mask
+           << std::dec << dendl;
+  ceph_assert((offset & mask) == 0);
+  ceph_assert((length & mask) == 0);
+  sbmap->set(offset >> order, length >> order);
+}
+
+static void clear_free_in_simple_bmap(CephContext* cct,
+  SimpleBitmap* sbmap, uint64_t offset, uint64_t length,
+  uint64_t mask, uint8_t order)
+{
+  dout(30) << "0x" << std::hex
+           << offset << "~" << length
+           << " " << mask
+           << std::dec << dendl;
+  ceph_assert((offset & mask) == 0);
+  ceph_assert((length & mask) == 0);
+  sbmap->clr(offset >> order, length >> order);
+}
+
+//-----------------------------------------------------------------------------------
+static void copy_simple_bitmap_to_allocator(SimpleBitmap* sbmap, Allocator* dest_alloc, uint64_t alloc_size)
+{
+  int alloc_size_shift = std::countr_zero(alloc_size);
+  uint64_t offset = 0;
+  extent_t ext    = sbmap->get_next_set_extent(offset);
+  while (ext.length != 0) {
+    dest_alloc->init_add_free(ext.offset << alloc_size_shift, ext.length << alloc_size_shift);
+    offset = ext.offset + ext.length;
+    ext = sbmap->get_next_set_extent(offset);
+  }
+}
+
+//-----------------------------------------------------------------------------------
+static void write_out_allocator_trailer(CephContext* cct,
+  BlueFS::FileWriter *p_handle, allocator_image_trailer* trailer)
+{
+  bufferlist trailer_bl;
+  encode(*trailer, trailer_bl);
+  uint32_t crc = -1;
+  crc = trailer_bl.crc32c(crc);
+  encode(crc, trailer_bl);
+  p_handle->append(trailer_bl);
+
+  dout(5) << "WRITE-entry_count=" << trailer->entries_count
+          << ", allocation_size=" << trailer->allocation_size
+          << ", format=" << trailer->format_version
+          << ", serial=" << trailer->serial
+          << dendl;
+}
+
+//-----------------------------------------------------------------------------------
+static int read_check_allocator_trailer(CephContext* cct,
+  BlueFS* bluefs, BlueFS::FileReader *p_handle, uint64_t offset,
+  const allocator_image_header *header, int trailer_size,
+  uint64_t extent_count, uint64_t read_alloc_size)
+{
+  ceph_assert(bluefs);
+  ceph_assert(p_handle);
+  ceph_assert(header);
+  {
+    bufferlist trailer_bl,temp_bl;
+    int        read_bytes = bluefs->read(p_handle, offset, trailer_size, &temp_bl, nullptr);
+    if (read_bytes != trailer_size) {
+      derr << "Failed bluefs->read() for trailer::read_bytes=" << read_bytes << ", req_bytes=" << trailer_size << dendl;
+      return -1;
+    }
+    offset += read_bytes;
+
+    trailer_bl.claim_append(temp_bl);
```

Comment:
Read directly into trailer_bl?

Reply:
BlueFS::_read invalidated output buffer prior to reading into it